Presence `last_update_id` API

mateuszmandera opened the pull request:
Implements a prototype for the API discussed in https://chat.zulip.org/#narrow/stream/378-api-design/topic/presence.20rewrite/near/1586819

Commit 1 Adds the necessary `last_update_id` column and `PresenceSequence` model, commit message elaborates on the migration plan
Commit 2 The backend+frontend parts of the system using the above data structure.

This is WIP, so no tests and only tested in the webapp.
What this has:
- Adds the `last_update_id` param and plumbing in the API through the backend and frontend
- Implements what is intended to be an efficient locking mechanism inside `do_update_user_presence`, updating `PresenceSequence` and the related `UserPresence` update in a single SQL query, at the very end of the transaction, ensuring the hot lock only gets held as briefly as possible.
- Verified the core mechanism seems to work; the webapp only fetches the recently updated presence data and updates its buddy list with the incremental info

What this does NOT have:
- [x] Doesn't handle the discussed case of getting presence info about a user the webapp doesn't know yet (https://chat.zulip.org/#narrow/stream/378-api-design/topic/presence.20rewrite/near/1586828) EDIT: Fixed
- [ ] Doesn't remove the `user_presence` query worker, which we probably want to do, moving presence updates in-process
- [x] `PresenceSequence` for the realm doesn't get created in the import codepath. EDIT: Fixed

mateuszmandera commented on zerver/actions/presence.py:
```diff
@@ -162,8 +175,65 @@ def do_update_user_presence(
+    #         presence.last_update_id = new_last_update_id
+    #         presence.save()
+    #     elif len(update_fields) > 0:
+    #         presence.last_update_id = new_last_update_id
+    #         presence.save(update_fields=[*update_fields, "last_update_id"])
+    #     presence_sequence.save(update_fields=["last_update_id"])
+    # But let's do it in a single, direct SQL query instead.
+
+    if created or len(update_fields) > 0:
+        query = sql.SQL("""
+            WITH new_last_update_id AS (
+                UPDATE zerver_presencesequence
+                SET last_update_id = last_update_id + 1
+                WHERE realm_id = {realm_id}
+                RETURNING last_update_id
+            )
+        """).format(realm_id=sql.Literal(user_profile.realm_id))
+
+        if created:
+            query += sql.SQL("""
+                INSERT INTO zerver_userpresence (user_profile_id, last_active_time, last_connected_time, realm_id, last_update_id)
+                VALUES ({user_profile_id}, {last_active_time}, {last_connected_time}, {realm_id}, (SELECT last_update_id FROM new_last_update_id))
+            """).format(
+                user_profile_id=sql.Literal(user_profile.id),
+                last_active_time=sql.Literal(presence.last_active_time),
+                last_connected_time=sql.Literal(presence.last_connected_time),
+                realm_id=sql.Literal(user_profile.realm_id),
+            )
+        else:
+            assert len(update_fields) > 0
+            update_fields_segment = sql.SQL(", ").join(
+                sql.SQL("{field} = {value}  ").format(
+                    field=sql.Identifier(field), value=sql.Literal(getattr(presence, field))
+                )
+                for field in update_fields
+            )
+            query += sql.SQL("""
+                UPDATE zerver_userpresence
+                SET {update_fields}, last_update_id = (SELECT last_update_id FROM new_last_update_id)
+                WHERE id = {presence_id}
```
This is the locking in `do_update_user_presence`, flagging it here for clarity. Idea is we do the `UPDATE` to the `PresenceSequence` in a CTE in the same query as the `UserPresence` update, fetching and using the incremented `last_update_id` for the `UserPresence.last_update_id`. The `UPDATE` is what acquires the hot lock.

Here are example queries generate by this:
- Creation of presence:

```sql
            WITH new_last_update_id AS (
                UPDATE zerver_presencesequence
                SET last_update_id = last_update_id + 1
                WHERE realm_id = 2
                RETURNING last_update_id
            )
        
                INSERT INTO zerver_userpresence (user_profile_id, last_active_time, last_connected_time, realm_id, last_update_id)
                VALUES (6, NULL, '2024-05-07T21:22:28+00:00'::timestamptz, 2, (SELECT last_update_id FROM new_last_update_id))

```

- Update:
    
    ```sql
                WITH new_last_update_id AS (
                    UPDATE zerver_presencesequence
                    SET last_update_id = last_update_id + 1
                    WHERE realm_id = 2
                    RETURNING last_update_id
                )
            
                    UPDATE zerver_userpresence
                    SET "last_connected_time" = '2024-05-07T21:22:30+00:00'::timestamptz, "last_active_time" = '2024-05-07T21:22:30+00:00'::timestamptz, last_update_id = (SELECT last_update_id FROM new_last_update_id)
                    WHERE id = 17
    ```

timabbott replied:
This looks good to me but maybe `SET {update_fields}` should be `SET {update_fields_segment}` rather than renaming it in the `.format` below, just for readability.

mateuszmandera replied:
This is done. Leaving the thread unresolved since it's useful highlighting this important locking mechanism for anyone else looking at the PR, with some context in my original comment above.

mateuszmandera commented on web/src/presence.ts:
```diff
@@ -177,7 +184,8 @@ export function set_info(
         }
     */
 
-    clear_internal_data();
+    presence_last_update_id = last_update_id;
```
Flagging this change, since it's small but important. `clean_internal_data()` purges all the presence data the webapp has, so that doesn't work with getting only recent presences from the server.

Seems to me that it's fine to just delete this call, so the webapp will keep the data it already has and patch the entries in it for the users that  got updated.

mateuszmandera replied:
But I don't understand why this call had to be here to begin with, because if we were fetching all data then the loop below would have just overwritten everything anyway. So maybe I'm missing something.

timabbott replied:
Digging with `git log -S'user_info = {}'  --follow -- static/js/activity.js`, I'm pretty sure the answer is there's no real reason; 5d4b2305fe0aee979fb241cd3b53e022ab107ff3 is what introduced this, and it looks like basically just an artifact of past refactoring.

timabbott commented on zerver/migrations/0520_user_presence_backfill_last_update_id_to_0.py:
```diff
@@ -0,0 +1,39 @@
+    BATCH_SIZE = 10000
+    lower_bound = 0
+
+    while lower_bound < max_id:
```
As noted in previous migrations, we should run past `max_id` by a `BATCH_SIZE` to deal with the possibility of races.

mateuszmandera replied:
Should we have some kind of semgrep rule for this then? These migrations are typically written with the idea that they're supposed to run after the server already loaded the new backend code that ensures new objects get created correctly; so running past `Max` is not needed under that assumption, so it feels pretty unintuitive to it. We have some other recent migrations that don't run past the `Max`, so we're not successfully enforcing this margin of error without some lint rule it seems.

mateuszmandera replied:
Tweaked both migrations to do the `max_id += BATCH_SIZE / 2` trick. Had to also update the `PresenceSequence` migration to filter out realms that already have the object:
```
        realm_ids = Realm.objects.filter(
            id__gt=lower_bound,
            id__lte=lower_bound + BATCH_SIZE,
            # Filter to realm whose PresenceSequence does not exist, to avoid
            # running into IntegrityError by trying to create duplicate PresenceSequence.
            presencesequence=None,
        ).values_list("id", flat=True)
```

timabbott replied:
In most cases, the failure mode of a race is just that the next migration (to enforce the backfill has completed) fails, so it's generally "safe" in that respect. But yeah, that seems like a good approach.

timabbott commented on zerver/migrations/0521_set_userpresence_last_update_id_not_null.py:
```diff
@@ -0,0 +1,17 @@
+            model_name="userpresence",
+            name="last_update_id",
+            field=models.PositiveBigIntegerField(db_index=True, default=0),
+        ),
```
This operation could probably be in migration 0520, since it's non-atomic anyway?

timabbott commented on zerver/migrations/0520_user_presence_backfill_last_update_id_to_0.py:
```diff
@@ -0,0 +1,39 @@
+    lower_bound = 0
+
+    while lower_bound < max_id:
+        UserPresence.objects.filter(id__gt=lower_bound, id__lte=lower_bound + BATCH_SIZE).update(
```
Might as well include `last_update_id=None` in the filter parameters.

timabbott commented on zerver/models/presence.py:
```diff
@@ -30,6 +30,8 @@ class UserPresence(models.Model):
     # queries to fetch all presence data for a given realm.
     realm = models.ForeignKey(Realm, on_delete=CASCADE)
 
+    last_update_id = models.PositiveBigIntegerField(db_index=True, default=0)
```
Probably worth a comment documenting this one. Maybe:

> The sequence ID within this realm for the last update to this user's presence; these IDs are generated by the `PresenceSequence` table and an important part of how we send incremental presence updates efficiently.

mateuszmandera replied:
Added the comments in both places.

timabbott commented on zerver/views/presence.py:
```diff
@@ -146,6 +146,7 @@ def update_active_status_backend(
     ping_only: Json[bool] = False,
     new_user_input: Json[bool] = False,
     slim_presence: Json[bool] = False,
+    presence_last_update_id: Optional[Json[int]] = None,
```
This is probably worth some discussion on what to call it; I'm thinking we want something to better hint that we're asking for "everything newer than this ID".

mateuszmandera replied:
https://chat.zulip.org/#narrow/stream/378-api-design/topic/presence.20rewrite/near/1797342

timabbott commented on zerver/lib/presence.py:
```diff
@@ -147,14 +147,24 @@ def get_presence_for_user(
         realm_id=realm.id,
+        # TODO: Should we remove this check when last_update_id
+        # is present, even if =-1?
         last_connected_time__gte=two_weeks_ago,
```
Yeah, I think we can do
```
    if last_update_id is not None:
        kwargs["last_update_id__gt"] = last_update_id
    else:
        kwargs["last_connected_time__gte"] = two_weeks_ago
```

mateuszmandera replied:
@timabbott Hmm I still wonder about this one. We nay want clients who use this api to send `last_update_id=-1` at startup to get all data, rather than not passing the param to do `last_update_id=None`; to allow the server to know that it's dealing a `last_update_id`-utilizing client. Even if right now we're not actually changed any behavior based on it.

So the question this brings is - should such a modern client initializing with `last_update_id=-1` get ALL presence data or just the data with the `last_connected_time__gte=two_weeks_ago` filter?

Of course if we want the two weeks limit on client initialization, we could just do 
```
    if last_update_id not in [None, -1]:
        kwargs["last_update_id__gt"] = last_update_id
    else:
        kwargs["last_connected_time__gte"] = two_weeks_ago
```

timabbott replied:
I think what we have is pretty reasonable -- seems like `last_update_id=-1` should be a good way to ask for "no really, give me all the data". But worth flagging this detail for API design discussion.

mateuszmandera replied:
https://chat.zulip.org/#narrow/stream/378-api-design/topic/presence.20rewrite.20.3E.20how.20much.20data.20to.20fetch.20at.20initialization

timabbott commented on zerver/lib/presence.py:
```diff
@@ -172,26 +182,47 @@ def get_presence_dict_by_realm(
 
 def get_presence_response(
-    requesting_user_profile: UserProfile, slim_presence: bool
+    requesting_user_profile: UserProfile, slim_presence: bool, last_update_id: Optional[int] = None
```
I feel like we probably want to call this `last_update_id` parameter something like `already_fetched_last_update_id` in the internal code.

mateuszmandera replied:
Did some renaming around to have these internal code variables named `last_update_id_fetched_by_client` and `last_update_id_fetched_by_server`

timabbott commented on zerver/actions/presence.py:
```diff
@@ -162,8 +175,65 @@ def do_update_user_presence(
+
+    # It's time to determine last_update_id and update the presence object in the database.
+    # This briefly takes the crucial lock on the PresenceSequence row for the user's realm.
+    # We're doing this in a single SQL query to avoid any Python-level overhead. We're also
```
Avoiding `Python-level overhead` is less the point than avoiding a network round-trip between the application server and the database while holding the lock.

mateuszmandera replied:
Hmm yeah, was thinking of this as part of the "Python-level overhead", but that's not the best phrasing. Updated to
`to avoid any unnecessary overhead, in particular database round-trips,`

timabbott commented on zerver/actions/presence.py:
```diff
@@ -109,10 +116,16 @@ def do_update_user_presence(
+        # so we defer that until the last moment.
+        # Create the presence object in-memory only for now.
+        presence = UserPresence(**defaults, user_profile=user_profile)
+        created = True
```
Maybe `creating` would be slightly more clear?

timabbott commented on web/src/activity.ts:
```diff
@@ -21,6 +21,7 @@ const post_presence_response_schema = z.object({
             }),
         )
         .optional(),
+    presence_last_update_id: z.number().optional(),
```
I think we don't need to tag this as optional, if we're passing the parameters that should make it always present.

mateuszmandera replied:
We need this optional, just like `presences` just above is `optional()`, because in the `ping_only: true` path, the server doesn't return these params.

And for that reason, in `send_presence_to_server()` below you have this block for handling the optionality in `ping_only: true` vs `false:
```
            if (redraw) {
                assert(
                    data.presences !== undefined,
                    "Presences should be present if not a ping only presence request",
                );
                assert(
                    data.server_timestamp !== undefined,
                    "Server timestamp should be present if not a ping only presence request",
                );
                assert(
                    data.presence_last_update_id !== undefined,
                    "Presence last update id should be present if not a ping only presence request",
                );

                presence.set_info(
                    data.presences,
                    data.server_timestamp,
                    data.presence_last_update_id,
                );
                redraw();
            }
```

timabbott replied:
Ahh, yes, sounds good. This kind of detail can be good to record in a comment.

mateuszmandera replied:
Added a comment in an extra commit.

mateuszmandera commented on zerver/views/events_register.py:
```diff
@@ -47,6 +47,7 @@ def events_register_backend(
         "client_gravatar", default=None, json_validator=check_bool
     ),
     slim_presence: bool = REQ(default=False, json_validator=check_bool),
+    presence_last_update_id: Optional[int] = REQ(default=None, json_validator=check_int),
```
Hmm yeah the idea was in case the mobile app wanted to use this; and also in case we were changing presence events format based on whether the client is `last_update_id`-based. But we're not changing events right now, so no point perhaps. Removed this for now.

timabbott replied:
I think a feature to have mobile be able to just used cached data on this detail is probably a good idea, so worth bringing it up in API design.

mateuszmandera replied:
https://chat.zulip.org/#narrow/stream/378-api-design/topic/presence.20rewrite.20.3E.20add.20last_update_to.20.2Fregister.3F

mateuszmandera commented:
Responded to comments and pushed a commit with a proposal for handling the main race:

> the discussed case of getting presence info about a user the webapp doesn't know yet (https://chat.zulip.org/#narrow/stream/378-api-design/topic/presence.20rewrite/near/1586828)

timabbott commented on web/src/presence.ts:
```diff
@@ -31,9 +31,11 @@ export type PresenceInfoFromEvent = {
+export const raw_info = new Map<number, RawPresence>();
 export const presence_info = new Map<number, PresenceStatus>();
 
+export const raw_info_for_unknown_users = new Map<number, RawPresence>();
```
Why do we need a special store? I would have thought that we could just store the information in the main `presence_info` / `raw_info` stores, and just have display code only iterate through users that it knows above via the main `people.ts` structures -- I thought that's roughly what we did in `main` already...

mateuszmandera replied:
Yeah, this approach works too after all, we can add the unknown user to the data, just making the buddy list doesn't trip up on it and that we redraw the user upon receiving the user creation event if the presence data already raced before it.

mateuszmandera commented:
This is becoming fairly complete, with now just some frontend test coverage missing, so removing the `WIP` and `prototype` claims.

mateuszmandera commented:
This now has test coverage and is passing CI. The commits are currently structured as a sequence of refinements on top of the main commit to keep things reviewable in logical chunks; feel free to squash if it's easier to review as one huge commit though.

timabbott commented on zerver/migrations/0526_presencesequence.py:
```diff
@@ -0,0 +1,73 @@
+
+        PresenceSequence.objects.bulk_create(
+            PresenceSequence(realm_id=realm_id, last_update_id=0) for realm_id in realm_ids
+        )
```
Do we need to include an `on_conflict` specification, in case another process adds a presence sequence row while this is running?

mateuszmandera replied:
There shouldn't be any way for that to happen. The code ensures to create `PresenceSequence` whenever it's creating a new realm, but doesn't touch old realms. Old realms are assumed in the later commit to just correctly have it (via the backfill migrations).

In particular, the `do_update_user_presence` logic intentionally doesn't try to create `PresenceSequence` for a realm if it's somehow missing - it would complicate reasoning about the queries and locking situation there.

`create_realm` creates the `PresenceSequence` in `transaction_atomic()` together with the `Realm` row creation, so there should also be no way for this migration to grab a realm in a not-yet-complete state either.

timabbott replied:
OK, that makes sense. But what this does mean is that presence API requests will 500 until this migration is complete... or maybe most correctly their queue may backlog? We can work this out in the deployment planning thread, which will include how to rework the commits in this PR.

timabbott commented on zerver/models/presence.py:
```diff
@@ -70,6 +76,11 @@ def status_from_string(status: str) -> Optional[int]:
 
+class PresenceSequence(models.Model):
+    realm = models.OneToOneField(Realm, on_delete=CASCADE)
+    last_update_id = models.PositiveBigIntegerField()
```
Should this have an explicit `default=0`?

mateuszmandera replied:
No strong preference, but I felt that code that creates `PresenceSequence`s should be mindful of what it's doing and set the value explicitly, so I left this without a `default`. We can add it easily too if that seems better though.

timabbott replied:
Yeah, OK, that reasoning seems fair.

timabbott commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,24 @@ format used by the Zulip server that they are interacting with.
 
 ## Changes in Zulip 9.0
 
+**Feature level 263**:
```
You need a blank line or the Markdown processor will misformat this.

timabbott commented on zerver/tests/test_events.py:
```diff
@@ -426,6 +426,11 @@ def normalize(state: Dict[str, Any]) -> None:
+                # We don't adjust presence_last_update_id via apply_events,
+                # since events don't carry the relevant information.
+                # Fix the value just like server_timestamp.
+                state["presence_last_update_id"] = 0
```
We should spell out in the comment that we pick 0, which a correct client should consider a noop, since `max(0, previous_presence_last_update_id) = previous_presence_last_update_id`

mateuszmandera replied:
Hmm I'm not sure I like adding such a comment. We don't set it to `0` because `max(0, previous_presence_last_update_id) = previous_presence_last_update_id`. It's just like `server_timestamp`, we set it to `0` because this is a "this value doesn't matter for the purposes of these tests". I don't think this should be interpreted under the frame of "what would this value mean to a client"? (`server_timestamp=0` also wouldn't be a meaningful value to a client; if anything, it's be just wrong).

And actually a client should be confused by seeing `presence_last_update_id=0` in almost any scenario - e.g. if there's no new presence data in the server response, this value will be equal to the `presence_last_update_id` it sent, not to `0`.

So as a reader, I'd find this kind of comment confusing due to it suggesting me to interpret a value that's not  actually meant to mean anything?

timabbott commented on zerver/lib/import_realm.py:
```diff
@@ -1072,6 +1076,12 @@ def do_import_realm(import_dir: Path, subdomain: str, processes: int = 1) -> Rea
+            re_map_foreign_keys(data, "zerver_presencesequence", "realm", related_table="realm")
+            bulk_import_model(data, PresenceSequence)
+        else:
+            PresenceSequence.objects.create(realm=realm, last_update_id=0)
```
This could probably benefit from a comment noting that we're enforcing the invariant that every realm has a `PresenceSequence` object.

timabbott commented on web/src/presence.ts:
```diff
@@ -198,19 +198,14 @@ export function set_info(
+        // GC'ed, and we're about to reload).
+        // Despite that, we still add the presence data to our structures,
+        // and it is the job of the code using them to correctly
+        // ignore these until the user becomes know.
```
"until we receive the basic metadata on this user."

timabbott commented on zerver/models/presence.py:
```diff
@@ -30,6 +30,13 @@ class UserPresence(models.Model):
+    # of how we send incremental presence updates efficiently.
+    # To put it simply, every time we update a UserPresence row in a realm,
+    # the row gets last_update_id equal to that of the previously
+    # updated row + 1.
```
"equal to 1 more than the previously updated row in that realm." -- just want to spell out explicitly the realm-specific nature of it, since one could easily imagine the opposite design.

timabbott commented on zerver/lib/presence.py:
```diff
@@ -172,26 +183,59 @@ def get_presence_dict_by_realm(
+        # If there are no results, that means that are no new updates to presence
+        # since what the client has last seen. Therefore, returning the same
+        # last_update_id that the client provided is correct.
+        last_update_id_fetched_by_server = last_update_id_fetched_by_client
```
We should probably discuss the "echo client" and "possibly returns `None`" details in the API design conversation -- I'm not entirely sure this protocol handles those corner cases in the best possible way.

mateuszmandera replied:
https://chat.zulip.org/#narrow/stream/378-api-design/topic/presence.20rewrite.20.3E.20response.20last_update_id.20when.20no.20new.20data

timabbott replied:
Looks great!

timabbott commented on zerver/lib/presence.py:
```diff
@@ -147,14 +147,25 @@ def get_presence_for_user(
+    else:
+        # TODO: How should clients set last_update_id_fetched_by_client at initialization?
+        # Should they do last_update_id_fetched_by_client = -1 to get *ALL* data?
+        kwargs["last_connected_time__gte"] = two_weeks_ago
```
Yeah, I think `-1` is probably correct, if the client . Worth mentioning this detail in "api design", since we'll want to document it in the API docs for the main presence API endpoint.

As a side note, we should probably aim to delete or rework `docs/subsystems/presence.md` to mainly link to the API docs with some extra notes, following the completion of this project and documenting the modern presence API. But probably not worth getting distracted with that until we can merge this.

mateuszmandera replied:
https://chat.zulip.org/#narrow/stream/378-api-design/topic/presence.20rewrite.20.3E.20how.20much.20data.20to.20fetch.20at.20initialization

mateuszmandera replied:
Removed this block to have `last_connected_time__gte` inside the main query unconditionally with a note:
```
        # We can consider tweaking this value when last_update_id is being used,
        # to potentially fetch more data since such a client is expected to only
        # do it once and then only do small, incremental fetches.
        last_connected_time__gte=two_weeks_ago,
```

timabbott replied:
Looks good!

timabbott commented on zerver/tests/test_presence.py:
```diff
@@ -139,6 +197,99 @@ def test_invalid_presence(self) -> None:
+        # The server reflects the client's last_update_id back to it,
+        # so in this edge case it's a None.
+        self.assertEqual(json["presence_last_update_id"], None)
+        self.assertFalse(UserPresence.objects.exists())
```
I don't love the idea of this field having a type that isn't `integer` from the standpoint of what clients need to validate. Probably worth discussing in the API design thread, but I think more rational might be to just fetch the current latest sequence number in that case? Not sure, but it'd be nice if clients don't need to think about parsing a parameter that might be `None`.

mateuszmandera replied:
A client that submitted `presence_last_update=None` (meaning they didn't send the param in the request at all; because e.g. this is the mobile app doing the fetch via `/register`) also doesn't even look at this value, so this might not matter? 

https://chat.zulip.org/#narrow/stream/378-api-design/topic/presence.20rewrite.20.3E.20response.20last_update_id.20when.20no.20new.20data

mateuszmandera replied:
Updated this to to be `self.assertEqual(json["presence_last_update_id"], -1)` as discussed in the thread

gnprice commented on zerver/migrations/0524_userpresence_last_update_id.py:
```diff
@@ -0,0 +1,37 @@
+            name="last_update_id",
+            field=models.PositiveBigIntegerField(db_index=True, default=0, null=True),
+        ),
+        AddIndexConcurrently(
```
A small optimization is that it should be more efficient to do the backfill first, and this AddIndexConcurrently after that, rather than vice versa.

That way the backfill doesn't have to churn the index while it's updating all the rows; effectively it means we write the index just once, instead of writing it and then writing over the whole thing.

mateuszmandera replied:
Oh yeah that's a nice idea, I moved this operation to the end of the `user_presence_backfill_last_update_id_to_0.py` migration

mateuszmandera commented on zerver/actions/presence.py:
```diff
@@ -162,11 +178,72 @@ def do_update_user_presence(
+    # We're doing this in a single SQL query to avoid any unnecessary overhead, in particular
+    # database round-trips.
+    # We're also intentionally doing this at the very end of the function, at the last step
+    # before the transaction commits. This ensures the lock is held for the shortest
```
Sure, added a note there:
```
    # Note: The lock isn't acquired explicitly via something like SELECT FOR UPDATE,
    # but rather we rely on the UPDATE statement taking an implicit row lock.
```

mateuszmandera commented on zerver/actions/presence.py:
```diff
@@ -109,10 +119,16 @@ def do_update_user_presence(
-        user_profile=user_profile,
-        defaults=defaults,
-    )
+    try:
+        presence = UserPresence.objects.select_for_update().get(user_profile=user_profile)
+        creating = False
+    except UserPresence.DoesNotExist:
+        # We're not ready to write until we know the next last_update_id value.
+        # We don't want to hold the lock on PresenceSequence for too long,
+        # so we defer that until the last moment.
+        # Create the presence object in-memory only for now.
+        presence = UserPresence(**defaults, user_profile=user_profile)
+        creating = True
```
Hmm yeah, this should be extremely rare since it can only happen on the very first presence update for a user (and I guess completely impossible until we remove the queue worker, since that serializes this?) so I left it initially and missed to add a note on that here.

If there's a conflict then the two concurrent presence creations must be so close that it doesn't matter which one we pick? I wanted to use `DO NOTHING` for that reason, but then this code would emit an event even though no write happen. So how about `DO UPDATE` instead with this comment and TODO?
```
            # There's a small possibility of a race where a diffrent process may have
            # already created a row for this user. Given the extremely close timing
            # of these events, there's no clear reason to prefer one over the other,
            # so we just go with the most direct approach of DO UPDATE, so that the
            # last event emitted (the one coming from our process, since we're the slower one)
            # matches the created presence state.
            # TODO: Might be worth changing this to DO NOTHING instead with a bit of extra logic
            # to skip emitting an event in such a scenario.
```

mateuszmandera replied:
> so that the last event emitted (the one coming from our process, since we're the slower one)

I believe this is not entirely accurate, because being slower on the database query doesn't guarantee we'll also be slower on emitting the event; but it should be close-enough to true and the `TODO` can be done as a follow-up guaranteeing a proper resolution? (and this isn't an issue until the queue worker is removed)

timabbott replied:
Yeah, I think this is OK to work through as a follow-up.

mateuszmandera replied:
#30364

mateuszmandera commented on zerver/migrations/0525_user_presence_backfill_last_update_id_to_0.py:
```diff
@@ -0,0 +1,47 @@
+    lower_bound = 0
+
+    # Make sure we run past the end in case of new rows created while we run.
+    max_id += BATCH_SIZE / 2
```
Yup. Doing/not doing something like this was briefly discussed in https://github.com/zulip/zulip/pull/29999/#discussion_r1593240912

Ultimately there's nothing that guarantees we complete other than having the server start running code that guarantees new `UserPresence` is created with this field initialized + that the migration grabs all `UserPresence`s prior to that. This should already be guaranteed by the described deployment plan without needing this `+= BATCH_SIZE / 2` margin at all.

mateuszmandera replied:
So not sure what pattern is best to use here, given that strictly speaking no margin of safety should be necessary at all.

timabbott replied:
For the thread, I rewrote the comment to explain correctly what we did when merging this commit; the logic seems like it'll work in practice, and importantly, the comment now details the safety trade-offs here, and in particular that the migration will fail-loud here.

timabbott commented:
I opened https://github.com/zulip/zulip/pull/30283 to integrate the migrations part of this, with the following changes.

``` diff
diff --git a/zerver/migrations/0526_user_presence_backfill_last_update_id_to_0.py b/zerver/migrations/0526_user_presence_backfill_last_update_id_to_0.py
index 7083776df6..babf436730 100644
--- a/zerver/migrations/0526_user_presence_backfill_last_update_id_to_0.py
+++ b/zerver/migrations/0526_user_presence_backfill_last_update_id_to_0.py
@@ -17,12 +17,14 @@ def backfill_user_presence_last_update_id(
     BATCH_SIZE = 10000
     lower_bound = 0
 
-    # Make sure we run past the end in case of new rows created while we run.
+    # Add a slop factor to make it likely we run past the end in case
+    # of new rows created while we run. The next step will fail to
+    # remove the null possibility if we race, so this is safe.
     max_id += BATCH_SIZE / 2
 
     while lower_bound < max_id:
         UserPresence.objects.filter(
-            id__gt=lower_bound, id__lte=lower_bound + BATCH_SIZE, last_update=None
+            id__gt=lower_bound, id__lte=lower_bound + BATCH_SIZE, last_update_id=None
         ).update(last_update_id=0)
         lower_bound += BATCH_SIZE
 
diff --git a/zerver/migrations/0527_presencesequence.py b/zerver/migrations/0527_presencesequence.py
index 222fd3b9c0..76a172c06c 100644
--- a/zerver/migrations/0527_presencesequence.py
+++ b/zerver/migrations/0527_presencesequence.py
@@ -20,7 +20,10 @@ def create_presence_sequence_for_old_realms(
     BATCH_SIZE = 2000
     lower_bound = 0
 
-    # Make sure we run past the end in case of new rows created while we run.
+    # Add a slop factor to make it likely we run past the end in case
+    # of new rows created while we run. Races with realm creation are
+    # pretty unlikely, and should throw an exception, so we should
+    # catch them.
     max_id += BATCH_SIZE / 2
 
     while lower_bound < max_id:
```

alexmv commented:
The first commit says `This builds on top of the previous commit` which is no longer the case, since the commit in question was merged in #30283, and there have been other things merged since then.

timabbott commented on web/src/buddy_data.ts:
```diff
@@ -312,10 +311,9 @@ function filter_user_ids(user_filter_text: string, user_ids: number[]): number[]
+        const person = people.maybe_get_user_by_id(user_id, true);
 
         if (!person) {
-            blueslip.warn("Got user_id in presence but not people: " + user_id);
```
Maybe a comment of the form:

// See the comments in presence.set_info for details, but this is an expected race. User IDs for whom we have presence but no user metadata should be skipped.

mateuszmandera commented:
Squashed and split thing into two deployable commits - first just backend API support, and then the frontend actually using it.

> The first commit says `This builds on top of the previous commit` which is no longer the case, since the commit in question was merged in #30283, and there have been other things merged since then.

Thanks, fixed that message to refer that commit specifically by its hash.

mateuszmandera commented on api_docs/changelog.md:
```diff
@@ -22,15 +22,19 @@ format used by the Zulip server that they are interacting with.
 
-* `POST /users/me/presence`: When `slim_presence` is used, a new
-  `presence_last_update_id` parameter can be given, instructing
+* `POST /users/me/presence`: A new `last_update_id`
```
This was a small bug in this doc btw, the param is called `last_update_id` for this endpoint. (the prefix would be redundant here, since obviously it's in the `presence` domain so to speak; there isn't other non-presence data included here that would require such disambiguation via prefixes)

mateuszmandera commented:
Pushed a commit tweaking the details for `last_update_id`+`slim_presence` interactions a bit, to fit https://chat.zulip.org/#narrow/stream/378-api-design/topic/presence.20rewrite/near/1817372 It should be squashed into the backed implementation commit, just leaving it separate to easily show what's changed.

timabbott commented:
Squashed the backend/API commits to merge as #30343.

timabbott commented:
Probably we can actually merge the rest of this too, I'll come back to that tomorrow.

timabbott commented:
Merged, thanks @mateuszmandera!
